torch/csrc/distributed/spmd/event_handler_impl.cpp

```diff
@@ -0,0 +1,209 @@
+    case EventType::PREPARE_MODULE: {
+      return handlePrepareModule(
+          c10::static_intrusive_pointer_cast<PrepareModuleEvent>(event));
+    }
+    case EventType::LOCAL_GRAD_READY: {
+      return handleLocalGradReady(
+          c10::static_intrusive_pointer_cast<LocalGradReadyEvent>(event));
+    }
+    case EventType::COMM_DONE: {
+      return {};
+    }
+    default:
+      TORCH_INTERNAL_ASSERT(
+          false,
+          "DefaultBucketer unexcepted event type",
+          event->schema().type_);
+  }
+}
+
+std::vector<std::shared_ptr<Future>> DefaultBucketer::handlePrepareModule(
+    c10::intrusive_ptr<PrepareModuleEvent> event) {
+  params_ = event->parameters();
+  return {};
+}
+
+std::vector<std::shared_ptr<Future>> DefaultBucketer::handleLocalGradReady(
+    c10::intrusive_ptr<LocalGradReadyEvent> event) {
+  std::vector<size_t> paramIndices;
+  paramIndices.reserve(1);
+  paramIndices.push_back(event->index());
+  auto bucket = std::make_shared<Bucket>(
+      event->index(), event->grad(), std::move(paramIndices));
+  return createOneFutureEvent<BucketReadyEvent>(
+      c10::make_intrusive<BucketReadyEvent>(std::move(bucket)));
+}
+
+std::vector<std::shared_ptr<Future>> DefaultBucketer::handleCommDone(
+    c10::intrusive_ptr<CommDoneEvent> event) {
+  const auto& bucket = event->bucket();
+  auto& grad = params_[bucket->index_].mutable_grad();
+  if (bucket->tensor_.data_ptr() != grad.data_ptr()) {
+    grad.copy_(bucket->tensor_, /*non_blocking=*/true);
+  }
+
+  // FIXME
+  TORCH_INTERNAL_ASSERT(bucket->paramIndices_.size() == 1);
+  return createOneFutureEvent<GlobalGradReadyEvent>(
+      c10::make_intrusive<GlobalGradReadyEvent>(bucket->paramIndices_[0]));
+}
+
+/////////////////////////////////////////////////////////////////////
+//                         AllReduceComm                           //
+/////////////////////////////////////////////////////////////////////
+
+std::vector<EventSchema> AllReduceComm::ingressEvents() const {
+  return {EventType::BUCKET_READY};
+}
+
+std::vector<EventSchema> AllReduceComm::egressEvents() const {
+  return {EventType::COMM_DONE};
+}
+
+std::vector<std::shared_ptr<Future>> AllReduceComm::handleEvent(
+    const c10::intrusive_ptr<Event>& event) {
+  switch (event->schema().type_) {
+    case EventType::BUCKET_READY: {
+      return handleBucketReady(
+          c10::static_intrusive_pointer_cast<BucketReadyEvent>(event));
+    }
+    default:
+      TORCH_INTERNAL_ASSERT(
+          false, "AllReduceComm unexcepted event type ", event->schema().type_);
+  }
+}
+
+std::vector<std::shared_ptr<Future>> AllReduceComm::handleBucketReady(
+    c10::intrusive_ptr<BucketReadyEvent> event) {
+  const auto& bucket = event->bucket();
+  std::vector<at::Tensor> buffers;
+  buffers.reserve(1);
+  buffers.push_back(bucket->tensor_);
+  pg_->allreduce(buffers)->wait();
+  return createOneFutureEvent<CommDoneEvent>(
+      c10::make_intrusive<CommDoneEvent>(bucket));
```

Comment:
nit: shouldn't we return a Future here that is retrieved from the allreduce? I guess we need the future/work merge.

Reply:
> I guess we need the future/work merge.

Yep, trying to avoid making this PR depend on future/work merge.

Reply:
Can you add a TODO to remind deprecating this temporary function once future/work merge is done?

Reply:
sure will do.